app/utils/communication/notifications.py

```diff
@@ -0,0 +1,361 @@
+            )
+
+            for device in firebase_devices:
+                firebase_tokens.append(device.firebase_device_token)
+
+        await self._add_message_for_user_in_database(
+            message=message,
+            tokens=firebase_tokens,
+            db=db,
+        )
+
+        try:
+            self._send_firebase_trigger_notification_by_topic(topic=topic)
+        except Exception as error:
+            hyperion_error_logger.warning(
+                f"Notification: Unable to send firebase notification for topic {topic}: {error}"
+            )
+
+    async def subscribe_user_to_topic(
+        self, topic: Topic, user_id: str, db: AsyncSession
+    ) -> None:
+        """
+        Subscribe a list of tokens to a given topic.
+        """
+        # Get all firebase_device_token related to the user
+        firebase_devices = await cruds_notification.get_firebase_devices_by_user_id(
+            user_id=user_id, db=db
+        )
+
+        firebase_device_tokens = [
+            device.firebase_device_token for device in firebase_devices
+        ]
+
+        self.subscribe_tokens_to_topic(tokens=firebase_device_tokens, topic=topic)
+
+        topic_membership = models_notification.TopicMembership(
+            user_id=user_id, topic=topic
+        )
+        await cruds_notification.create_topic_membership(
+            topic_membership=topic_membership, db=db
+        )
+
+    async def unsubscribe_user_to_topic(
+        self, topic: Topic, user_id: str, db: AsyncSession
+    ) -> None:
+        """
+        Subscribe a list of tokens to a given topic.
+        """
+        # Get all firebase_device_token related to the user
+        firebase_devices = await cruds_notification.get_firebase_devices_by_user_id(
+            user_id=user_id, db=db
+        )
+
+        firebase_device_tokens = [
+            device.firebase_device_token for device in firebase_devices
+        ]
+
+        self.unsubscribe_tokens_to_topic(tokens=firebase_device_tokens, topic=topic)
+
+        await cruds_notification.delete_topic_membership(
+            topic=topic, user_id=user_id, db=db
+        )
+
+
+class NotificationTool:
+    """
+    Utility class to send notifications in the background.
+
+    This class should be instantiated before each use with a `BackgroundTasks` manager.
+    The best way to do so would be to use a
+    """
+
+    def __init__(
+        self,
+        background_tasks: BackgroundTasks,
+        notification_manager: NotificationManager,
+        db: AsyncSession,
+    ):
+        self.background_tasks = background_tasks
+        self.notification_manager = notification_manager
+        self.db = db
+
+    async def send_notification_to_user(self, user_id: str, message: Message):
+        await self.notification_manager.send_notification_to_user(
+            user_id=user_id,
+            message=message,
+            db=self.db,
+        )
+        # self.background_tasks.add_task(
```

Comment:
You can delete comments

Reply:
Thanks for pointing this part out. I decommented the code as sending notification as background service is the wanted behaviour, now everything else works.